pkg/processor/procbuiltin/unwraprecord.go

```diff
@@ -0,0 +1,171 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procbuiltin
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/foundation/multierror"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/record"
+)
+
+const (
+	unwrapRecordPayloadProcType = "unwraprecord"
+
+	// todo: support openCDC format
+	unwrapRecordConfigFormat = "format" // only debezium for now
+)
+
+func init() {
+	processor.GlobalBuilderRegistry.MustRegister(unwrapRecordPayloadProcType, UnwrapRecordPayload)
+}
+
+// UnwrapRecordPayload unwraps the record format from the payload and returns the record as openCDC format
+func UnwrapRecordPayload(config processor.Config) (processor.Interface, error) {
+	return unwrapRecord(unwrapRecordPayloadProcType, recordPayloadGetSetter{}, config)
+}
+
+func unwrapRecord(
+	processorType string,
+	getSetter recordDataGetSetter,
+	config processor.Config,
+) (processor.Interface, error) {
+	var (
+		err    error
+		format string
+		r      record.Record
+	)
+
+	format = config.Settings[unwrapRecordConfigFormat]
+	if format == "" {
+		return nil, cerrors.Errorf("%s: record format type not specified", processorType)
+	}
+
+	return processor.InterfaceFunc(func(_ context.Context, rec record.Record) (record.Record, error) {
+		data := getSetter.Get(rec) // record.payload.after
+		var jsonData record.StructuredData
+		switch d := data.(type) {
+		case record.RawData:
+			// unmarshal raw data to structured
+			err = json.Unmarshal(data.Bytes(), &jsonData)
+			if err != nil {
+				return record.Record{}, cerrors.Errorf("%s: %w", processorType, err)
+			}
+		case record.StructuredData:
+			jsonData = d
+		default:
+			return record.Record{}, cerrors.Errorf("%s: unexpected data type %T", processorType, data)
+		}
+
+		payloadJSON, ok := jsonData[DebeziumFieldPayload]
+		if !ok {
+			return record.Record{}, cerrors.Errorf("%s: the payload is missing from the record", processorType)
+		}
+
+		// cast payload into a map
+		payload := payloadJSON.(map[string]any)
+		// validate all the debezium fields exist
+		err = checkDebeziumFieldsExist(payload)
+		if err != nil {
+			return record.Record{}, cerrors.Errorf("%s: %w", processorType, err)
+		}
+
+		// convert json data into a record
+		before := convertJSONToStructuredData(payload[DebeziumFieldBefore])
+		after := convertJSONToStructuredData(payload[DebeziumFieldAfter])
+		metadata := convertJSONToMetadata(payload[DebeziumFieldSource])
+
+		r.Payload.Before = before
+		r.Payload.After = after
+		r.Metadata = metadata
+		r.Operation = getRecordOp(payload[DebeziumFieldOp].(string))
```

Comment:
just leaving a note here from our pairing:
- need to include the `r.Position`
- need to include the `r.Key`, pending the approach we decide on for the key